### PR TITLE
forbid `#[track_caller]` on main

### LINKED ISSUE #75125: Program crashes with #[track_caller] applied to main()
### Code

```Rust
#[track_caller]
fn main() {
    panic!("{}: oh no", std::panic::Location::caller());
}
```

### Meta

Crash occurs in both 1.46.0 beta and in current nightly.

`rustc --version --verbose`:
```
rustc 1.46.0-beta.2 (6f959902b 2020-07-23)
binary: rustc
commit-hash: 6f959902b3103c49ca981fbc01871589c3498489
commit-date: 2020-07-23
host: x86_64-unknown-linux-gnu
release: 1.46.0-beta.2
LLVM version: 10.0
```

```
rustc 1.47.0-nightly (81e754c35 2020-08-02)
binary: rustc
commit-hash: 81e754c359c471f91263813c46c67955071716a7
commit-date: 2020-08-02
host: x86_64-unknown-linux-gnu
release: 1.47.0-nightly
LLVM version: 10.0

```

### Error output

```
memory allocation of 94396265198519 bytes failed
```

### PR DESCRIPTION
**Pull request by lcnr**
fixes #75125

cc @anp

### CODE REVIEW

**Review by anp — APPROVED** [review]

File: src/librustc_typeck/lib.rs [hunk 1]
```diff
@@ -274,6 +291,23 @@ fn check_start_fn_ty(tcx: TyCtxt<'_>, start_def_id: LocalDefId) {
                         .emit();
                         error = true;
                     }
+
+                    for attr in it.attrs {
+                        if attr.check_name(sym::track_caller) {
+                            tcx.sess
+                                .struct_span_err(
+                                    attr.span,
+                                    "start is not allowed to be `#[track_caller]`",
```
Comment (ecstatic-morse) [inline src/librustc_typeck/lib.rs:300]:
```suggestion
                                    "`start` is not allowed to be `#[track_caller]`",
```

Same below. Also, maybe this should just say "program entry point is not allowed..."? Up to you.

**Comment by ecstatic-morse** [comment]
r=me with the error message tweaked.

File: src/librustc_typeck/lib.rs (hunk 1 above)
Comment (lcnr) [reply src/librustc_typeck/lib.rs]: Went with `start` for now

**Comment by lcnr** [comment]
come on :cry: I blame vs code here

Thanks for noticing @bors r=ecstatic-morse

I should look into automatically preventing this, considering how often this has happend to me

**Comment by JohnTitor** [comment]
The failure is spurious: https://github.com/rust-lang/cargo/issues/8517
@bors retry

**Comment by ehuss** [comment]
@lcnr Would you mind adding a small note to https://github.com/rust-lang/reference/blob/master/src/attributes/codegen.md#the-track_caller-attribute about this restriction?

**Comment by lcnr** [comment]
not exactly sure how to best formulate this but opened https://github.com/rust-lang/reference/pull/872 for now

**Comment by spastorino** [comment]
Adding missing `T-compiler` label so this is picked up by prioritization automation bot and placed into the compiler team meeting agenda.

**Comment by spastorino** [comment]
[discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/21643weeklymeeting2020080554818.html#206146290) in T-compiler meeting.

@rustbot modify labels: beta-accepted

**Pull request merged**